File: backend/app/core/roles.py
"""
Role hierarchy and permission service — Single Responsibility Principle.

This module is the ONLY place where role logic lives.
All role checks across the app delegate here.

Hierarchy:
    super_admin  →  system-wide god mode
    org_admin    →  organization owner (Project Manager)
    manager      →  team leader
    employee     →  regular worker

Invite rules:
    org_admin  can invite  →  manager, employee
    manager    can invite  →  employee
    employee   can invite  →  nobody
"""

from enum import IntEnum


class Role(IntEnum):
    """
    Roles ordered by power level (higher = more privileged).
    Using IntEnum lets us compare roles with simple < > operators.
    """

    EMPLOYEE = 0
    MANAGER = 1
    ORG_ADMIN = 2
    SUPER_ADMIN = 3

    @classmethod
    def from_str(cls, value: str) -> "Role":
        """Convert a DB string like 'org_admin' to the enum."""
        mapping = {
            "employee": cls.EMPLOYEE,
            "manager": cls.MANAGER,
            "org_admin": cls.ORG_ADMIN,
            "super_admin": cls.SUPER_ADMIN,
        }
        role = mapping.get(value)
        if role is None:
            raise ValueError(f"Unknown role: {value}")
        return role

    def to_str(self) -> str:
        """Convert the enum back to the DB string."""
        return {
            Role.EMPLOYEE: "employee",
            Role.MANAGER: "manager",
            Role.ORG_ADMIN: "org_admin",
            Role.SUPER_ADMIN: "super_admin",
        }[self]

    @property
    def display_name(self) -> str:
        """Human-readable label for the UI."""
        return {
            Role.EMPLOYEE: "Employee",
            Role.MANAGER: "Manager",
            Role.ORG_ADMIN: "Admin",
            Role.SUPER_ADMIN: "Super Admin",
        }[self]


# ── Invite rules ──────────────────────────────────────────────
# Which roles each role is allowed to create/invite.

_INVITE_PERMISSIONS: dict[Role, frozenset[Role]] = {
    Role.SUPER_ADMIN: frozenset({Role.ORG_ADMIN, Role.MANAGER, Role.EMPLOYEE}),
    Role.ORG_ADMIN: frozenset({Role.MANAGER, Role.EMPLOYEE}),
    Role.MANAGER: frozenset({Role.EMPLOYEE}),
    Role.EMPLOYEE: frozenset(),
}


class RoleService:
    """
    Stateless service — all methods are pure functions.
    Open/Closed: extend by adding to the enums/dicts above, not by modifying methods.
    """

    # ── Hierarchy queries ──

    @staticmethod
    def is_at_least(user_role: str, minimum_role: str) -> bool:
        """Check if user_role >= minimum_role in the hierarchy."""
        return Role.from_str(user_role) >= Role.from_str(minimum_role)

    @staticmethod
    def is_above(user_role: str, other_role: str) -> bool:
        """Check if user_role is strictly above other_role."""
        return Role.from_str(user_role) > Role.from_str(other_role)

    # ── Invite checks ──

    @staticmethod
    def can_invite(inviter_role: str, target_role: str) -> bool:
        """Can a user with inviter_role invite someone with target_role?"""
        inviter = Role.from_str(inviter_role)
        target = Role.from_str(target_role)
        return target in _INVITE_PERMISSIONS.get(inviter, frozenset())

    @staticmethod
    def invitable_roles(inviter_role: str) -> list[str]:
        """Return the list of role strings this inviter is allowed to assign."""
        inviter = Role.from_str(inviter_role)
        return [r.to_str() for r in sorted(_INVITE_PERMISSIONS.get(inviter, frozenset()))]

    # ── Feature access ──

    @staticmethod
    def can_view_team_activity(role: str) -> bool:
        return Role.from_str(role) >= Role.MANAGER

    @staticmethod
    def can_view_all_screenshots(role: str) -> bool:
        return Role.from_str(role) >= Role.MANAGER

    @staticmethod
    def can_manage_projects(role: str) -> bool:
        return Role.from_str(role) >= Role.ORG_ADMIN

    @staticmethod
    def can_manage_users(role: str) -> bool:
        """Admin can manage all users; manager can manage employees."""
        return Role.from_str(role) >= Role.MANAGER

    @staticmethod
    def can_view_reports(role: str) -> bool:
        return Role.from_str(role) >= Role.MANAGER

    @staticmethod
    def can_change_role(changer_role: str, target_current_role: str, target_new_role: str) -> bool:
        """
        Can `changer_role` change someone from `target_current_role` to `target_new_role`?
        Rules:
          - You must be strictly above BOTH the current and new role.
          - The new role must be in your invite permissions.
        """
        changer = Role.from_str(changer_role)
        current = Role.from_str(target_current_role)
        new = Role.from_str(target_new_role)
        allowed = _INVITE_PERMISSIONS.get(changer, frozenset())
        return changer > current and new in allowed


# Module-level singleton for convenience
role_service = RoleService()

File: frontend/src/lib/roles.ts
/**
 * Role hierarchy and permission utilities — frontend mirror of backend RoleService.
 * Single source of truth for all role checks in the UI.
 *
 * Hierarchy (higher number = more power):
 *   employee (0) → manager (1) → org_admin (2) → super_admin (3)
 */

export const ROLES = {
  employee: 0,
  manager: 1,
  org_admin: 2,
  super_admin: 3,
} as const;

export type RoleKey = keyof typeof ROLES;

export const ROLE_LABELS: Record<RoleKey, string> = {
  employee: "Employee",
  manager: "Manager",
  org_admin: "Admin",
  super_admin: "Super Admin",
};

export const ROLE_COLORS: Record<RoleKey, string> = {
  employee: "bg-slate-100 text-slate-700",
  manager: "bg-blue-100 text-blue-700",
  org_admin: "bg-purple-100 text-purple-700",
  super_admin: "bg-red-100 text-red-700",
};

/** Check if userRole >= minimumRole in the hierarchy */
export function isAtLeast(userRole: string, minimumRole: RoleKey): boolean {
  return (ROLES[userRole as RoleKey] ?? -1) >= ROLES[minimumRole];
}

/** Check if userRole is strictly above otherRole */
export function isAbove(userRole: string, otherRole: string): boolean {
  return (ROLES[userRole as RoleKey] ?? -1) > (ROLES[otherRole as RoleKey] ?? -1);
}

/** Sidebar items each role can see */
export function canViewNav(userRole: string, navKey: string): boolean {
  const role = userRole as RoleKey;
  const power = ROLES[role] ?? 0;

  switch (navKey) {
    // Everyone sees these
    case "dashboard":
    case "time-tracking":
    case "settings":
      return true;

    // Manager+ sees these
    case "reports":
    case "activity":
    case "screenshots":
      return power >= ROLES.manager;

    // Manager+ sees Team (they can add employees)
    case "team":
      return power >= ROLES.manager;

    // Admin+ sees Projects
    case "projects":
      return power >= ROLES.org_admin;

    default:
      return true;
  }
}

/** Roles this user can assign when inviting */
export function invitableRoles(userRole: string): RoleKey[] {
  switch (userRole) {
    case "super_admin":
      return ["org_admin", "manager", "employee"];
    case "org_admin":
      return ["manager", "employee"];
    case "manager":
      return ["employee"];
    default:
      return [];
  }
}

/** Can this user invite anyone at all? */
export function canInvite(userRole: string): boolean {
  return invitableRoles(userRole).length > 0;
}
